web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java

```diff
@@ -0,0 +1,64 @@
+package org.mobilitydata.gtfsvalidator.web.service.util;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.File;
+import java.util.Base64;
+import org.mobilitydata.gtfsvalidator.input.CountryCode;
+import org.mobilitydata.gtfsvalidator.runner.ValidationRunner;
+import org.mobilitydata.gtfsvalidator.runner.ValidationRunnerConfig;
+import org.mobilitydata.gtfsvalidator.util.VersionResolver;
+import org.mobilitydata.gtfsvalidator.web.service.controller.GoogleCloudPubsubMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for validating GTFS feeds. */
+public class ValidationHandler {
+  private final Logger logger = LoggerFactory.getLogger(ValidationHandler.class);
+
+  /**
+   * Validates the GTFS feed zip file, and stores the results in a local temp directory using the
+   * job ID as the directory name.
+   *
+   * @param feedFile
+   * @param jobId
+   * @param countryCode
+   * @return the path to the temp directory containing the validation results
+   */
+  public File validateFeed(File feedFile, String jobId, String countryCode) {
+    var runner = new ValidationRunner(new VersionResolver());
+    var tempDir = feedFile.getParentFile();
+    var outputPath = new File(tempDir.toPath() + jobId);
```

Comment:
Maybe the outputPath should just be constructed in the caller, so that validateFeed doesn't need to be aware of jobId nor does it need to return the outputPath?

Reply:
@bdferris-v2 this makes sense to me, though I tried to pass in outputPath and move lines 30, 31 to the calling method to derive it, and now I get a compilation error:
```
Note: .../gtfs-validator/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
```
Any advice would be appreciated, I'm not sure how to fix.

Reply:
I'm not sure what your change looks like, but the following compiles for me:

```
diff --git a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
index a25aa498..5b4059cb 100644
--- a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
+++ b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
@@ -103,9 +103,11 @@ public class ValidationController {
 
       // copy the file from GCS to a temp directory
       File tempFile = storageHelper.createTempFile(jobId, fileName);
+      File outputPath = tempFile.getParentFile();
+
       // extracts feed files from zip to temp output directory, validates, and returns
       // the path to the output directory
-      File outputPath = handler.validateFeed(tempFile, jobId, countryCode);
+      handler.validateFeed(tempFile, outputPath, countryCode);
 
       // upload the extracted files and the validation results from outputPath to GCS
       storageHelper.uploadFilesToStorage(jobId, outputPath);
diff --git a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
index 581b4d87..8b6e53c1 100644
--- a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
+++ b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
@@ -25,10 +25,8 @@ public class ValidationHandler {
    * @param countryCode
    * @return the path to the temp directory containing the validation results
    */
-  public File validateFeed(File feedFile, String jobId, String countryCode) {
+  public File validateFeed(File feedFile, File outputPath, String countryCode) {
     var runner = new ValidationRunner(new VersionResolver());
-    var tempDir = feedFile.getParentFile();
-    var outputPath = new File(tempDir.toPath() + jobId);
     var configBuilder =
         ValidationRunnerConfig.builder()
             .setGtfsSource(feedFile.toURI())
```

Reply:
I think that's exactly what I tried, but I will try again to duplicate.

Reply:
@bdferris-v2 So, we need to have a different output path and delete the tempDir at the end (preserving what it was doing before), so this is my change:

```
diff --git a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
index eb09602c..83d8c929 100644
--- a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
+++ b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/controller/ValidationController.java
@@ -93,9 +93,12 @@ public class ValidationController {
 
       // copy the file from GCS to a temp directory
       File tempFile = storageHelper.copyFromStorageToTempFile(jobId, fileName);
+
+      tempDir = tempFile.getParentFile();
+      var outputPath = new File(tempDir.toPath() + jobId);
       // extracts feed files from zip to temp output directory, validates, and returns
       // the path to the output directory
-      File outputPath = handler.validateFeed(tempFile, jobId, countryCode);
+      handler.validateFeed(tempFile, outputPath, countryCode);
 
       // upload the extracted files and the validation results from outputPath to GCS
       storageHelper.uploadFilesToStorage(jobId, outputPath);
diff --git a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
index c5f5cae2..a7029ada 100644
--- a/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
+++ b/web/service/src/main/java/org/mobilitydata/gtfsvalidator/web/service/util/ValidationHandler.java
@@ -21,10 +21,8 @@ public class ValidationHandler {
    * @param countryCode
    * @return the path to the temp directory containing the validation results
    */
-  public File validateFeed(File feedFile, String jobId, String countryCode) {
+  public File validateFeed(File feedFile, File outputPath, String countryCode) {
     var runner = new ValidationRunner(new VersionResolver());
-    var tempDir = feedFile.getParentFile();
-    var outputPath = new File(tempDir.toPath() + jobId);
     var configBuilder =
         ValidationRunnerConfig.builder()
             .setGtfsSource(feedFile.toURI())
```
and when I build, I see the following:

```

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':aggregateJavadoc'.
> Javadoc generation failed. Generated Javadoc options file (useful for troubleshooting): '/Users/brian/development/jarvus/gtfs-validator/build/tmp/aggregateJavadoc/javadoc.options'

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
24 actionable tasks: 11 executed, 13 up-to-date
```

Full build output [here](https://gist.github.com/briandonahue/b85fe593550fe8b3fc62c6c7d0d511ed)

Reply:
🤦 in my defense, it outputs a LOT of info, I totally missed that! 😬